contracts/RewardsAggregator.sol

```diff
@@ -0,0 +1,262 @@
+        application = application_;
+        rewardsHolder = rewardsHolder_;
+        oldCumulativeMerkleDrop = _oldCumulativeMerkleDrop;
+    }
+
+    function setMerkleRoot(bytes32 merkleRoot_) external override onlyOwner {
+        emit MerkleRootUpdated(merkleRoot, merkleRoot_);
+        merkleRoot = merkleRoot_;
+    }
+
+    function setRewardsHolder(address rewardsHolder_) external onlyOwner {
+        require(
+            rewardsHolder_ != address(0),
+            "Rewards Holder must be an address"
+        );
+        emit RewardsHolderUpdated(rewardsHolder, rewardsHolder_);
+        rewardsHolder = rewardsHolder_;
+    }
+
+    /**
+     * @notice Returns the amount of rewards that a given stake has already
+     *         claimed from the Merkle distribution, including the old Merkle
+     *         distribution contract. The returned amount does not include the
+     *         claimed rewards generated by the applications.
+     */
+    function cumulativeMerkleClaimed(
+        address stakingProvider
+    ) public view returns (uint256) {
+        uint256 newAmount = cumulativeClaimed[stakingProvider];
+        if (newAmount > 0) {
+            return newAmount;
+        } else {
+            return oldCumulativeMerkleDrop.cumulativeClaimed(stakingProvider);
+        }
+    }
+
+    /**
+     * @notice Claim the rewards that have been generated by the Merkle
+     *         distribution mechanism.
+     */
+    function claimMerkle(
+        address stakingProvider,
+        address beneficiary,
+        uint256 cumulativeAmount,
+        bytes32 expectedMerkleRoot,
+        bytes32[] calldata merkleProof
+    ) public {
+        require(merkleRoot == expectedMerkleRoot, "Merkle root was updated");
+
+        // Verify the merkle proof
+        bytes32 leaf = keccak256(
+            abi.encodePacked(stakingProvider, beneficiary, cumulativeAmount)
+        );
+        require(
+            verifyMerkleProof(merkleProof, expectedMerkleRoot, leaf),
+            "Invalid proof"
+        );
+
+        // Mark it claimed (potentially taking into consideration state in old
+        // MerkleDistribution contract)
+        uint256 preclaimed = cumulativeMerkleClaimed(stakingProvider);
+        require(preclaimed < cumulativeAmount, "Nothing to claim");
+        cumulativeClaimed[stakingProvider] = cumulativeAmount;
+
+        // Send the tokens
+        unchecked {
+            uint256 amount = cumulativeAmount - preclaimed;
+            IERC20(token).safeTransferFrom(rewardsHolder, beneficiary, amount);
+            emit MerkleClaimed(
+                stakingProvider,
+                amount,
+                beneficiary,
+                expectedMerkleRoot
+            );
+        }
+    }
+
+    /**
+     * @notice Check if a particular stake has available rewards to claim.
+     */
+    function canClaimApps(address stakingProvider) public view returns (bool) {
+        return application.availableRewards(stakingProvider) > 0;
```

Comment:
@vzotova IIRC, is it possible that this method produces outdated results in case of penalties, right? I'm wondering what can be the effect.

Reply:
this method uses penalty calculation which is actual, so even if penalty is over but reset was not done - it's still correct method to call